### ISSUE TITLE
JSON validation errors should return a clear error to users

### ISSUE THREAD

**Issue by rhyolight** [numenta/nupic#1717]
Invalid JSON in the swarm description results in an error like this:

```
Results from all experiments:
----------------------------------------------------------------
Generating experiment files in directory: /tmp/tmpwVsELA...
Writing  313 lines...
Writing  113 lines...
done.
None
json.loads(jobInfo.results) raised an exception.  Here is some info to help with debugging:
jobInfo:  _jobInfoNamedTuple(jobId=1006, client=u'GRP', clientInfo=u'', clientKey=u'', cmdLine=u'$HYPERSEARCH', params=u'{"hsVersion": "v2", "maxModels": null, "persistentJobGUID": "351cf396-8d38-11e4-a734-685d43b983b8", "useTerminators": false, "description": {"includedFields": [{"fieldName": "timestamp", "fieldType": "datetime"}, {"maxValue": 53.0, "fieldName": "kw_energy_consumption", "fieldType": "float", "minValue": 0.0}], "streamDef": {"info": "kw_energy_consumption", "version": 1, "streams": [{"info": "Rec Center", "source": "file://rec-center-hourly.csv", "columns": ["*"]}]}, "inferenceType": "TemporalMultiStep", "inferenceArgs": {"predictionSteps": [1], "predictedField": "kw_energy_consumption"}, "iterationCount": -1, "swarmSize": "medium"}}', jobHash='5\x1c\xfa\x9e\x8d8\x11\xe4\xa74h]C\xb9\x83\xb8', status=u'notStarted', completionReason=None, completionMsg=None, workerCompletionReason=u'success', workerCompletionMsg=None, cancel=0, startTime=None, endTime=None, results=None, engJobType=u'hypersearch', minimumWorkers=1, maximumWorkers=4, priority=0, engAllocateNewWorkers=1, engUntendedDeadWorkers=0, numFailedWorkers=0, lastFailedWorkerErrorMsg=None, engCleaningStatus=u'notdone', genBaseDescription=None, genPermutations=None, engLastUpdateTime=datetime.datetime(2014, 12, 26, 19, 48, 47), engCjmConnId=None, engWorkerState=None,
engStatus=None, engModelMilestones=None)
jobInfo.results:  None
EXCEPTION:  expected string or buffer
Traceback (most recent call last):
  File "swarm.py", line 109, in <module>
    swarm(INPUT_FILE)
  File "swarm.py", line 101, in swarm
    modelParams = swarmForBestModelParams(SWARM_DESCRIPTION, name)
  File "swarm.py", line 78, in swarmForBestModelParams
    verbosity=0
  File "/usr/lib/python2.7/site-packages/nupic/swarming/permutations_runner.py", line 276, in runWithConfig
    return _runAction(runOptions)
  File "/usr/lib/python2.7/site-packages/nupic/swarming/permutations_runner.py", line 217, in _runAction
    returnValue = _runHyperSearch(runOptions)
  File "/usr/lib/python2.7/site-packages/nupic/swarming/permutations_runner.py", line 160, in _runHyperSearch
    metricsKeys=search.getDiscoveredMetricsKeys())
  File "/usr/lib/python2.7/site-packages/nupic/swarming/permutations_runner.py", line 825, in generateReport
    results = json.loads(jobInfo.results)
  File "/usr/lib/python2.7/site-packages/nupic/support/object_json.py", line 163, in loads
    json.loads(s, object_hook=objectDecoderHook, **kwargs))
  File "/usr/lib/python2.7/json/__init__.py", line 351, in loads
    return cls(encoding=encoding, **kw).decode(s)
  File "/usr/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
TypeError: expected string or buffer
```

We could do better by wrapping this and returning a more detailed error to the user that tells them to ensure their JSON is valid.

**Comment by tmczhangtielin** [comment]
I meet exactly this problem today, may you tell me what can I do to solve this problem?

**Comment by rhyolight** [comment]
@tmczhangtielin Your swarm_description.py file is not valid JSON. Try running through http://jsonlint.com/ to find out why.

**Comment by kalletammemae** [comment]
I got the same error although SWARM_CONFIG appeared to be perfect JSON. Error disappeared when I defined shell variable NUPIC assigned to Nupic source directory and exported it. Finally added export statement to .bashrc i.e. export NUPIC=<path to source directory>. It worked too when NUPIC was assigned just to home directory.

**Comment by rhyolight** [comment]
> Error disappeared when I defined shell variable NUPIC assigned to Nupic source directory and exported it. Finally added export statement to .bashrc i.e. export NUPIC=. It worked too when NUPIC was assigned just to home directory.

@kalletammemae That's strange. Can you tell me exactly what command you ran that created the error? Was this a script you wrote that did swarming yourself? If so, can you provide your own stacktrace and the line of code that started the error in your script? Many thanks.

**Comment by kalletammemae** [comment]
Well, indeed :-) It was Sine Wave prediction tutorial what I tried on new computer with new nupic binary installation. Running command sine-prediction I was stuck with JSON exception. Looking advice in web I found discussion here. Tested JSON in SWARM_CONFIG and it was fine.
I discovered NUPIC variable issue rather accidentally. Downloaded source and declared NUPIC variable just for case. The code was working in one terminal and in another not and only difference was that in existence of NUPIC in one. It was surprising how this one influences JSON parsing but experiment was repeatable.
